### PR TITLE
DOC: Refactor Numba enhancing performance and add parallelism caveat

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

### CODE REVIEW

**Comment by jreback** [comment]
> doc/source/user_guide/window.rst:264:12: F401 'pandas.api.indexers.BaseIndexer' imported but unused
doc/source/user_guide/window.rst:281:22: F821 undefined name 'CustomIndexer'
doc/source/user_guide/enhancingperf.rst:341:18: E225 missing whitespace around operator

**Review by jreback — CHANGES_REQUESTED** [review]
linting issue

File: doc/source/user_guide/enhancingperf.rst [hunk 1]
```diff
@@ -414,25 +437,20 @@ Consider the following toy example of doubling each observation:
...
 prefer that Numba throw an error if it cannot compile a function in a way that
 speeds up your code, pass Numba the argument
-``nopython=True`` (e.g.  ``@numba.jit(nopython=True)``). For more on
+``nopython=True`` (e.g.  ``@jit(nopython=True)``). For more on
 troubleshooting Numba modes, see the `Numba troubleshooting page
 <https://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#the-compiled-code-is-too-slow>`__.
 
-Read more in the `Numba docs <https://numba.pydata.org/>`__.
+Using ``parallel=True`` (e.g. ``@jit(parallel=True)``) may use an unsafe threading layer to execute parallelism, potentially
+leading to segfaults. You may need to first `specify a safe threading layer <https://numba.readthedocs.io/en/stable/user/threading-layer.html#selecting-a-threading-layer-for-safe-parallel-execution>`__
```
Comment (stuartarchibald) [inline doc/source/user_guide/enhancingperf.rst:452]: Numba should not end up causing a `SIGSEGV`, if this happens, it's probably a bug, please report it!

Numba might print a message to `STDERR` saying it's detected unsafe behaviour and then signal with `SIGABRT` to protect users from corruption that would occur in the case of user defined code violating the assumptions of certain threading layers.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for updating these docs. I've a query (inline) relating to `SIGSEGV`. If you are (or indeed anyone is!) finding segmentation faults arising in the use of Numba then please [report them](https://github.com/numba/numba/issues). Numba tries to ensure it will:
* not segfault on legal code.
* catch illegal code at compile time (i.e. before executing and segfaulting/corrupting etc)!
* catch illegal states in the threading layers and abort before corruption occurs.

Thanks again!

File: doc/source/user_guide/enhancingperf.rst (hunk 1 above)
Comment (mroeschke) [reply doc/source/user_guide/enhancingperf.rst]: Thanks for the clarification in behavior. I've clarified in this section that `SIGSEGV` are bugs that should be reported instead.

Comment (stuartarchibald) [reply doc/source/user_guide/enhancingperf.rst]: Thank you, much appreciated.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mroeschke

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**